app/cdap/components/hydrator/components/Canvas/index.tsx

```diff
@@ -0,0 +1,563 @@
+        position: {
+          x: parseInt(node._uiPosition.left, 10),
+          y: parseInt(node._uiPosition.top, 10),
+        },
+        selected: false,
+      };
+      if (shouldShowAlertsPort(node) && shouldShowErrorsPort(node)) {
+        reactflowNode.type = 'pluginWithAlertAndError';
+      }
+      if (existingIds.includes(node.name)) {
+        reactflowNode.position = reactNodes.find((nd) => nd.id === node.name).position;
+        reactflowNode.selected = reactNodes.find((nd) => nd.id === node.name).selected;
+      }
+      return reactflowNode;
+    });
+  };
+
+  const [nodes, setNodes, onNodesChange] = useNodesState([]);
+  const [edges, setEdges, onEdgesChange] = useEdgesState([]);
+
+  // draw connections
+  const onConnect = useCallback(
+    (params) => {
+      setEdges((eds) => addEdge(params, eds));
+    },
+    [setEdges]
+  );
+
+  const onSelectionChange = useCallback((elements) => {
+    setSelectedElements(elements);
+    console.log('Selected elements:', elements);
+  }, []);
+
+  const getSelectedNodesAndConnections = () => {
+    const selectedNodes = selectedElements.nodes;
+    const selectedEdges = selectedElements.edges;
+    const selectedNodesId = selectedNodes.map((node) => node.id);
+    const selectedEdgesId = selectedEdges.map((edge) => edge.id);
+    return { selectedNodesId, selectedEdgesId };
+  };
+
+  const copySelectedElements = () => {
+    setCopiedElements(selectedElements);
+  };
+
+  const copySelectedNodeId = (nodeId: string) => {
+    const node = findNodeWithNodeId(nodes, nodeId);
+    setCopiedElements({ nodes: [node], edges: [] });
+  };
+
+  const deleteSelectedElements = () => {
+    const { selectedNodesId, selectedEdgesId } = getSelectedNodesAndConnections();
+    setNodes((nds) => nds.filter((node) => !selectedNodesId.includes(node.id)));
+    setEdges((eds) => eds.filter((edge) => !selectedEdgesId.includes(edge.id)));
+    const newNodes = getAngularNodes().filter((node) => !selectedNodesId.includes(node.name));
+    updateNodes(newNodes);
+    const newConnections = getAngularConnections().filter(
+      (conn) =>
+        !selectedNodesId.includes(conn.from) &&
+        !selectedNodesId.includes(conn.to) &&
+        !selectedEdgesId.find((edge) => edge.includes(conn.from) && edge.includes(conn.to))
+    );
+    updateConnections(newConnections);
+    setSelectedElements({
+      nodes: [],
+      edges: [],
+    });
+  };
+
+  const deleteSelectedNodeId = (nodeId: string) => {
+    setNodes((nds) => nds.filter((node) => node.id !== nodeId));
+    const newNodes = getAngularNodes().filter((node) => node.name !== nodeId);
+    updateNodes(newNodes);
+    const newConnections = getAngularConnections().filter(
+      (conn) => nodeId !== conn.from && nodeId !== conn.to
+    );
+    updateConnections(newConnections);
+  };
+
+  const pasteCopiedElements = () => {
+    let pastedNodes = [...copiedElements.nodes];
+    let pastedEdges = [...copiedElements.edges];
```

Comment:
Can this interact with the Clipboard API? I.e., can we support copying in one browser window and paste in another?

Should the details of copy/paste be in the Canvas implementation, or closer to the state management?

Reply:
You are right. I overlooked this usage. Will make changes